hadoop-hdds/rocksdb-checkpoint-differ/src/main/java/org/apache/ozone/rocksdiff/CompactionLogEntry.java

```diff
@@ -0,0 +1,179 @@
+    // Encoding is used to deal with \n. Protobuf appends \n after each
+    // parameter. If ByteArray is simply converted to a string or
+    // protobuf.toString(), it will contain \n and will be added to the log.
+    // Which creates a problem while reading compaction logs.
+    // Compaction log lines are read sequentially assuming each line is one
+    // compaction log entry.
+    return Base64.getEncoder().encodeToString(getProtobuf().toByteArray());
+  }
+
+  public static CompactionLogEntry fromEncodedString(String string) {
+    try {
+      byte[] decode = Base64.getDecoder().decode(string);
+      return getFromProtobuf(CompactionLogEntryProto.parseFrom(decode));
+    } catch (InvalidProtocolBufferException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public String toString() {
+    return String.format("dbSequenceNumber: '%s', inputFileInfoList: '%s'," +
+            " outputFileInfoList: '%s',", dbSequenceNumber, inputFileInfoList,
+        outputFileInfoList);
+  }
+
+  public static CompactionLogEntry fromCompactionFiles(
+      long dbSequenceNumber,
+      List<String> inputFiles,
+      List<String> outputFiles
+  ) {
+
+    try (ManagedOptions options = new ManagedOptions();
+         ManagedReadOptions readOptions = new ManagedReadOptions()) {
+      List<FileInfo> inputFileInfos = convertFileInfo(inputFiles, options,
+          readOptions);
+      List<FileInfo> outputFileInfos = convertFileInfo(outputFiles, options,
+          readOptions);
+      return new CompactionLogEntry(dbSequenceNumber, inputFileInfos,
+          outputFileInfos);
+    }
+  }
+
+  private static List<FileInfo> convertFileInfo(
+      List<String> sstFiles,
+      ManagedOptions options,
+      ManagedReadOptions readOptions
+  ) {
+    if (CollectionUtils.isEmpty(sstFiles)) {
+      return Collections.emptyList();
+    }
+
+    final int fileNameOffset = sstFiles.get(0).lastIndexOf("/") + 1;
+    List<FileInfo> response = new ArrayList<>();
+
+    for (String sstFile : sstFiles) {
+      String fileName = sstFile.substring(fileNameOffset,
+          sstFile.length() - SST_FILE_EXTENSION_LENGTH);
+      SstFileReader fileReader = new SstFileReader(options);
+      try {
+        fileReader.open(sstFile);
+        String columnFamily = StringUtils.bytes2String(
+            fileReader.getTableProperties().getColumnFamilyName());
+        SstFileReaderIterator iterator = fileReader.newIterator(readOptions);
+        iterator.seekToFirst();
+        String startKey = StringUtils.bytes2String(iterator.key());
+        iterator.seekToLast();
+        String endKey = StringUtils.bytes2String(iterator.key());
+
+        FileInfo fileInfo = new FileInfo(fileName, startKey, endKey,
+            columnFamily);
+        response.add(fileInfo);
+      } catch (RocksDBException rocksDBException) {
+        throw new RuntimeException("Failed to read SST file: " + sstFile,
+            rocksDBException);
```

Comment:
What happens when `onCompactionCompleted` throws this? RocksDB compaction thread crash? OM crash?

Reply:
Based on Swami's finding: https://github.com/facebook/rocksdb/blob/1e63fc9925919257d0435ea206763acc6b92a1c6/include/rocksdb/listener.h#L560C1-L562C77
It could cause undefined behavior. We need to handle it properly. I'll update it.